lib/experimental/graph.nim

```diff
@@ -0,0 +1,428 @@
+## Generic implementation of graph data structure
+
+import std/[
+  tables,
+  strformat,
+  deques,
+  hashes,
+  options,
+  math,
+  lenientops,
+  strutils,
+  sequtils,
+  algorithm,
+  intsets,
+  sets
+]
+
+#[
+checklist
+
+- [ ] All implementations work both at compile and run-time
+- [ ] Works on JS target
+- [ ] All iterator algorithms have both mutable and immutable implementations
+- [ ] As little side effects as possible
+- [ ] Make node and edge ID's distinct integers to avoid messing them up
+- [ ] Iterateo over node value pairs (source, target: N, edge: E)
+
+- IDEA :: Add `->` and `<-` as operators that would return `TempEdge`
+  object that can be added via `graph.add source -> target`.
+
+- TODO :: Test if cycle detection works for graphs that contain overlapping
+  cycles (`0 -> 1 -> 0` and `1 -> 2 -> 1`)
+
+]#
+
+type
+  HGraphProperty* = enum
+    gpDirected
+    gpUndirected
+    gpAllowSelfLoops
+    gpIgnoreSelfLoops
+
+  HEdge*[Id] = tuple[src, dst: Id]
+  HGraph*[Id] = object
+    flags: set[HGraphProperty]
+    edges: OrderedTable[Id, seq[Id]]
+
+type
+  HGraphError* = ref object of CatchableError
+  HGraphCyclesError* = ref object of HGraphError
+
+func newHGraph*[Id](
+    properties: set[HGraphProperty] = {gpDirected, gpAllowSelfLoops}
+  ): HGraph[Id] =
+
+  HGraph[Id](flags: properties)
+
+func newHGraph*[Id](
+    values: openarray[(Id, Id)],
+    properties: set[HGraphProperty] = {gpDirected, gpAllowSelfLoops}
+  ): HGraph[Id] =
+  result = newHGraph[Id](properties)
+  for (src, dst) in values:
+    result.edges.mgetOrPut(src, @[]).add dst
+
+func default*[Id](graph: typedesc[HGraph[Id]]): HGraph[Id] =
+  newHGraph[Id]()
+
+func contains*[Id](graph: HGraph[Id], node: Id): bool =
+  node in graph.edges
+
+
+func contains*[Id](graph: HGraph[Id], edge: HEdge[Id]): bool =
+  result = (
+    ( edge.src in graph.edges and
+      edge.dst in graph.edges[edge.src] ) or
+    ( gpUndirected in graph.flags and
+      edge.dst in graph.edges and
+      edge.src in graph.edges[edge.dst] )
+  )
+
+func getEdge*[Id](graph: HGraph[Id], source, target: Id): HEdge[Id] =
+  if source.id in graph.structure.outgoingIndex:
+    for outEdge in graph.structure.outgoingIndex[source.id]:
+      if graph.structure.edgeMap[outEdge.id][1] == target:
+        return outEdge
+
+  raise newException(ValueError, "No edge for source/target pair")
+
+func edgeCount*[Id](graph: HGraph[Id]): int =
+  for node, targets in pairs(graph.edges):
+    result += targets.len()
+
+func nodeCount*[Id](graph: HGraph[Id]): int =
+  graph.edges.len()
+
+iterator nodes*[Id](graph: HGraph[Id]): Id =
+  for i, _ in pairs(graph.edges):
+    yield i
+
+proc allNodes*[Id](graph: HGraph[Id]): seq[Id] =
```

Comment:
Generally recommendation keep the lightweight accessors and traversals (iterators) up top. Then the "heavier" procs like `allNodes` following those. Finally the mutators and "scary" stuff following that.

Typically, when learning/first using a data structure module folks are:
1. How do I make one, easy mode? (unless they already have one then start at the next step)
2. I got one, now what can I do with it? (reads, accessors, traversals)
3. How do I put this inside?
4. OK now more advanced bits

Of course, dependencies between routines might not allow following this 'perfectly'